### ISSUE TITLE
Whether the gradients are computed right when training in multi-gpu

### ISSUE THREAD

**Issue by yuzhile** [ildoonet/tf-pose-estimation#16]
The gradients are computed at each towers in example of multi_gpu_cifar10 and then averaged. However there is no average operation in your code. How do you ensure the parameters can be update right?

**Comment by ildoonet** [comment]
1. I split input batch and everything here : https://github.com/ildoonet/tf-openpose/blob/master/train.py#L73

2. do forward pass with splitted batch on each gpus. : https://github.com/ildoonet/tf-openpose/blob/master/train.py#L82

3. When I compute the loss, all the informations are gathered : https://github.com/ildoonet/tf-openpose/blob/master/train.py#L110

4. Also, When gradients are computed, each gradients will be assigned to gpus which calculate the gradient and update the shared weights. : https://github.com/ildoonet/tf-openpose/blob/master/train.py#L129

**Comment by tpys** [comment]
Thanks for sharing your code, I am confused the multi GPU training too.  It seems that you sum loss and update them on the last GPU instead of CPU,  this way is different from tensorflow's multi GPU tutorial, they update all model parameters on the CPU because transferring data to and from GPUs is quite slow, which is shown below.  I am new to tensorflow, sorry for wrong understanding.

![](https://www.tensorflow.org/images/Parallelism.png)

**Comment by ildoonet** [comment]
@tpys 

Above diagram is also right. But it is not necessary to compute average gradients on CPU. Moving data from gpu to gpu is not that slow when your hardware have a SLI. 

Gradients on Summed Loss will be split into multiple tower(multiple gpu), so my code will train the model with gpus.

**Comment by tpys** [comment]
@ildoonet I got it, thank you very much.
